From b7eb8af9bc52b409251429a0f2b503c293a87e95 Mon Sep 17 00:00:00 2001
From: Joey Riches <josephriches@gmail.com>
Date: Tue, 18 Jul 2023 15:23:53 +0100
Subject: [PATCH] Solus: Ensure we don't enable rootless xorg on nvidia-drm

The wrapper code assumes that if a driver supports KMS, then rootless xorg
will work. Unfortunately with the nvidia-drm driver (the KMS equivalent
of nvidia) this doesn't seem to be the case.

True as of: Nvidia 535.54.03.
---
 hw/xfree86/xorg-wrapper.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/hw/xfree86/xorg-wrapper.c b/hw/xfree86/xorg-wrapper.c
index 6c1022358..276934eb2 100644
--- a/hw/xfree86/xorg-wrapper.c
+++ b/hw/xfree86/xorg-wrapper.c
@@ -231,6 +231,13 @@ int main(int argc, char *argv[])
     }
 
 #ifdef WITH_LIBDRM
+    char driver_name[64] = {0};
+    struct drm_version ver = {
+        .name = driver_name,
+        .name_len = 63,
+    };
+    int info;
+
     /* Detect if we need root rights, except when overridden by the config */
     if (needs_root_rights == -1) {
         for (i = 0; i < 16; i++) {
@@ -243,7 +250,11 @@ int main(int argc, char *argv[])
 
             memset(&res, 0, sizeof(struct drm_mode_card_res));
             r = ioctl(fd, DRM_IOCTL_MODE_GETRESOURCES, &res);
-            if (r == 0)
+
+            /* HACKS! rootless xorg doesn't work with nvidia-drm (tested 535.54.03) */
+            info = ioctl(fd, DRM_IOCTL_VERSION, &ver);
+
+            if (r == 0 && strncmp(driver_name, "nvidia-drm", strlen("nvidia-drm")) != 0)
                 kms_cards++;
 
             close(fd);
-- 
2.42.0

